File: backend/src/main/java/arbremojo/backend/model/Justification.java
package arbremojo.backend.model;

import arbremojo.backend.model.user.Student;
import jakarta.persistence.*;
import lombok.*;
import org.hibernate.annotations.ColumnDefault;

import java.util.List;

@Entity
@Table(name = "justification")
@Setter
@Getter
@ToString
@AllArgsConstructor
@NoArgsConstructor
public class Justification {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "justification_id")
    private Integer justificationId;

    @Column(name = "reason", nullable = false)
    private String reason;

    @Column(name = "accepted")
    private Boolean accepted;

    @Column(name = "student_list_id", nullable = false)
    private Integer studentListId;

    @Column(name = "student_id", nullable = false)
    private Integer studentId;

    @ManyToOne(fetch = FetchType.EAGER, targetEntity = Student.class)
    @JoinColumn(name = "student_id", insertable = false, updatable = false)
    private Student student;

    @OneToOne(fetch = FetchType.EAGER, targetEntity = StudentList.class)
    @JoinColumn(name = "student_list_id", insertable = false, updatable = false)
    private StudentList studentList;

    @OneToMany(fetch = FetchType.EAGER, targetEntity = JustificationImage.class)
    @JoinColumn(name = "justification_id", insertable = false, updatable = false)
    private List<JustificationImage> justificationImageList;
}

File: frontend/src/model/justification.ts
import {JustificationImage} from "./justification-image";
import {Student} from "./user/student";
import {StudentList} from "./student-list";

export class Justification {
  justificationId: number | undefined;
  reason: string;
  accepted: boolean | null | undefined;

  studentListId: number;
  studentId: number;

  justificationImageList: JustificationImage[] = [];

  student: Student | undefined;
  studentList: StudentList | undefined;

  constructor(reason: string, studentListId: number, studentId: number, accepted: boolean | null | undefined = null, justificationId?: number) {
    this.reason = reason;
    this.accepted = accepted;
    this.studentListId = studentListId;
    this.studentId = studentId;
    this.justificationId = justificationId;
  }

  static fromJson(json: Justification): Justification {
    let justification = new Justification(json.reason, json.studentListId, json.studentId, json.accepted, json.justificationId);

    justification.justificationImageList = JustificationImage.initializeJustificationImages(json.justificationImageList);
    if (json.student != undefined) justification.student = Student.fromJson(json.student!);
    if (json.studentList != undefined) justification.studentList = StudentList.fromJson(json.studentList!);

    return justification;
  }

  static initializeJustifications(json: Justification[]): Justification[] {
    let justifications: Justification[] = [];
    for (let i = 0; i < json.length; i++) {
      justifications.push(Justification.fromJson(json[i]));
    }
    return justifications;
  }
}
